src/uct/cuda/cuda_copy/cuda_copy_iface.c

```diff
@@ -517,15 +572,62 @@ static UCS_CLASS_INIT_FUNC(uct_cuda_copy_iface_t, uct_md_h md, uct_worker_h work
     return UCS_OK;
 }
 
+static void uct_cuda_copy_stream_destroy(CUstream *stream, int valid_ctx)
+{
+    if ((*stream == NULL) || (!valid_ctx)) {
+        return;
+    }
+
+    UCT_CUDADRV_FUNC_LOG_WARN(cuStreamDestroy(*stream));
+    *stream = NULL;
+}
+
+static void uct_cuda_copy_ctx_rsc_destroy(uct_cuda_copy_ctx_rsc_t *ctx_rsc)
+{
+    uct_cuda_copy_ctx_mpool_priv_t *mpool_priv;
+    int valid_ctx;
+    ucs_memory_type_t src, dst;
+    ucs_queue_head_t *event_q;
+
+    mpool_priv = ucs_mpool_priv(&ctx_rsc->cuda_event_desc);
+    valid_ctx  = uct_cuda_copy_ctx_rsc_valid(mpool_priv->validator);
+
+    ucs_memory_type_for_each(src) {
+        ucs_memory_type_for_each(dst) {
+            event_q = &ctx_rsc->queue_desc[src][dst].event_queue;
+            if (!ucs_queue_is_empty(event_q)) {
+                ucs_warn("stream destroyed but queue not empty");
+            }
+
+            uct_cuda_copy_stream_destroy(&ctx_rsc->queue_desc[src][dst].stream,
```

Comment:
to consider - IMO passing valid_ctx to uct_cuda_copy_stream_destroy is redundant - if context is destroyed we don't have to go over all streams in the first place
but on other hand need to warn if stream event q is not empty